mbuild/compound.py

```diff
@@ -991,7 +991,13 @@ def _to_topology(self, atom_list, chains=None, residues=None):
         for atom in atom_list:
             # Chains
             for parent in atom.ancestors():
-                if chains and parent.name in chains:
+                if chains and atom.name in chains:
```

Comment:
This should happen before we go into the ancestors. Only if the current particle does not match the name do we start looking up the hierarchy for any that do match.